pymc/distributions/timeseries.py

```diff
@@ -237,64 +298,13 @@ def logp(self, x):
         -------
         TensorVariable
         """
-        if x.ndim > 0:
-            x_im1 = x[:-1]
-            x_i = x[1:]
-            mu, sigma = self._mu_and_sigma(self.mu, self.sigma)
-            innov_like = Normal.dist(mu=x_im1 + mu, sigma=sigma).logp(x_i)
-            return self.init.logp(x[0]) + at.sum(innov_like)
-        return self.init.logp(x)
-
-    def random(self, point=None, size=None):
-        """Draw random values from GaussianRandomWalk.
-
-        Parameters
-        ----------
-        point: dict, optional
-            Dict of variable values on which random values are to be
-            conditioned (uses default point if not specified).
-        size: int, optional
-            Desired size of random sample (returns one sample if not
-            specified).
-
-        Returns
-        -------
-        array
-        """
-        # sigma, mu = distribution.draw_values([self.sigma, self.mu], point=point, size=size)
-        # return distribution.generate_samples(
-        #     self._random,
-        #     sigma=sigma,
-        #     mu=mu,
-        #     size=size,
-        #     dist_shape=self.shape,
-        #     not_broadcast_kwargs={"sample_shape": to_tuple(size)},
-        # )
-        pass
-
-    def _random(self, sigma, mu, size, sample_shape):
-        """Implement a Gaussian random walk as a cumulative sum of normals.
-        axis = len(size) - 1 denotes the axis along which cumulative sum would be calculated.
-        This might need to be corrected in future when issue #4010 is fixed.
-        """
-        if size[len(sample_shape)] == sample_shape:
-            axis = len(sample_shape)
-        else:
-            axis = len(size) - 1
-        rv = stats.norm(mu, sigma)
-        data = rv.rvs(size).cumsum(axis=axis)
-        data = np.array(data)
-
-        # the following lines center the random walk to start at the origin.
-        if len(data.shape) > 1:
-            for i in range(data.shape[0]):
-                data[i] = data[i] - data[i][0]
-        else:
-            data = data - data[0]
-        return data
+        rv, updates = aesara.scan(
```

Comment:
Alright heres my question, how do I properly implement a the logp method? Here's why I'm confused on this question

1. The [distribution docs](https://docs.pymc.io/en/latest/contributing/developer_guide_implementing_distribution.html#inheriting-from-a-pymc-base-distribution-class) and other distributions such as [Kumarswamy](https://github.com/pymc-devs/pymc/blob/main/pymc/distributions/continuous.py#L1359) show calculating the logp totally "manually"
2. Some distributions like [Exponential](https://github.com/pymc-devs/pymc/blob/main/pymc/distributions/continuous.py#L1420) don't have a logp at all, my guess is because the rv_op is defined in Aesara?
3. The pretty cool colab that Ricardo put together [uses scan](https://gist.github.com/ricardoV94/a4da531a60a0f16facf8dbb5416a2cd4#file-aeppl_scan-py-L22)

Out of the three approaches I'm not entirely sure which one should be used here, as well as some general confusion about the logp"less" distributions

Reply:
> Alright heres my question, how do I properly implement a the logp method? Here's why I'm confused on this question
> 
>     1. The [distribution docs](https://docs.pymc.io/en/latest/contributing/developer_guide_implementing_distribution.html#inheriting-from-a-pymc-base-distribution-class) and other distributions such as [Kumarswamy](https://github.com/pymc-devs/pymc/blob/main/pymc/distributions/continuous.py#L1359) show calculating the logp totally "manually"
> 
>     2. Some distributions like [Exponential](https://github.com/pymc-devs/pymc/blob/main/pymc/distributions/continuous.py#L1420) don't have a logp at all, my guess is because the rv_op is defined in Aesara?
> 
>     3. The pretty cool colab that Ricardo put together [uses scan](https://gist.github.com/ricardoV94/a4da531a60a0f16facf8dbb5416a2cd4#file-aeppl_scan-py-L22)
> 
> 
> Out of the three approaches I'm not entirely sure which one should be used here, as well as some general confusion about the logp"less" distributions

1. and 3. The manual implementation is the "standard" distribution way, which you wanted to try as the first step. The point of later switching to Aeppl is that it will compute it for you, so there is no need to write it by hand.
2.  Those are implemented in Aeppl here: https://github.com/aesara-devs/aeppl/blob/333108143a1f4b63d9fcd9842dbe35457025c180/aeppl/logprob.py#L191-L198 (we should probably leave a comment so that devs/users can find it more easily

Reply:
Here something like the following should be enough:

```python
innit_logp = pm.logp(pm.Normal.dist(mu, sigma), value[:1] - init)
innov_logp = pm.logp(pm.Normal.dist(mu, sigma), at.diff(value))
total_logp = at.concatenate([innit_logp, innov_logp])
return total_logp
```

No need for a scan for the case of GaussianRandomWalk logp

Reply:
You shouldn't call aeppl inside a logp expression, because aeppl actually needs the pairs `{random_variable: value_variable}` in the graph, but when `GaussianRandomWalk.logp` is called you only have the `value_variables`.

Reply:
> but when GaussianRandomWalk.logp is called it's only given the random_variables for this

Thanks for all the advice, I'm going give it a shot today.
Im confused on point, the docs show `logp` taking `value` and `param1`?
https://docs.pymc.io/en/latest/contributing/developer_guide_implementing_distribution.html
<img width="664" alt="image" src="https://user-images.githubusercontent.com/7213793/148655150-68fdd5af-7d90-4ec5-ae67-877bb4c9f298.png">